eeps/0005-many-variants.md

```diff
@@ -0,0 +1,224 @@
++  openssl_1_1 = openssl.v1_1;
++  openssl_3 = openssl.v3_0;
++  openssl_3_2 = openssl.v3_2;
++  openssl_3_3 = openssl.v3_3;
+```
+
+```nix
+# pkgs/openssl/variants.nix
+{
+  v1_1 = {
+    version = "1.1.1w";
+    hash = "sha256-zzCYlQy02FOtlcCEHx+cbT3BAtzPys1SHZOSUgi3asg=";
+  };
+  ...
+}
+```
+
+```nix
+# pkgs/openssl/default.nix
+{ callPackage
+, mkPolyPkg
+, ...
+}@args:
+
+# TODO: future proposal, make this less ugly
+callPackage (mkPolyPkg {
+  variants = ./variants.nix;
+  aliases = ./aliases.nix;
+  defaultSelector = (p: p.v3_3);
+  genericBuilder = ./generic.nix;
+}) args
+```
+
+New generic.nix:
+```nix
+# pkgs/openssl/generic.nix
+{ version
+, hash
+, packageOlder
+, packageAtLeast
+, packageBetween
+, mkVersionPassthru
+, ...
+}:
+
+# Former packaging arguments:
+{ lib
+, stdenv
+...
+}@args:
+
+stdenv.mkDerivation {
+  ...
+  passthru = (mkVersionPassthru args ) // {
+    ... # existing args
+  };
+}
+```
+
+Usage of openssl from top-level:
+```
+nix-repl> openssl     
+«derivation /nix/store/ayvv16slswqfps7c1zql8rml5nla6zky-openssl-3.3.1.drv»
+
+nix-repl> openssl.v3_2 
+«derivation /nix/store/77794w2h7xslv02k54xxc8x7krc1aazg-openssl-3.2.2.drv»
+
+# Overriding still works
+nix-repl> openssl.v3_2.override { withDocs = false; }
+«derivation /nix/store/2196b3bj4vk2p6i7f3ai3p0n1m5im5iz-openssl-3.2.2.drv»
+
+# To build all variants, need to pass `allowAliases = false;` to avoid eval error
+nix-build -A openssl.variants
+```
+
+## Unresolved issues
+
+- Argument passing is ugly
+  - This is caused by us trying to satisfy mkPolyPkg and the package's arguments with the same arguments
+  - This could be improved by importing default.nix with just mkPolyPkg, and passing the partially applied mkPolyPkg function to callPackage
+    - E.g. `openssl = callPackage (import ./pkgs/openssl { inherit mkPolyPkg; }) { };` 
+    - Similar to auto call-by-name, we could also create another directory which would just auto call generic packages, TBD in another proposal.
+
+```nix
+# pkgs/openssl/default.nix, with the result of this partially applied function
+# being passed to callPackage
+{ mkPolyPkg }:
+
+mkPolyPkg {
+  variants = ./variants.nix;
+  aliases = ./aliases.nix;
+  defaultSelector = (p: p.v3_3);
+  genericBuilder = ./generic.nix;
+}
+```
+
+- Should you be able to recurse variants? E.g. `openssl.v3_2.v3_1` (returns v3.1)
+- Add `.overrideVersion`?
+  - Unsure if having yet-another way to `.overrideX` is worth it
+  - In most cases, this should be the same as doing `drv.overrideAttrs { version = ...; src = ...; }`
+  - However, this would apply version information before the expression gets called with package arguments
+    - This allows for logic which references version in many locations to work without use of mkDerivation fixed points. E.g. `(finalAttrs: { ... })`
+
+## Future work
+
+- Mitigate argument passing awkwardness with mkAutoCalledGenericDir proposal (TBD).
+- Provide updater-script for use with generic packages
```

Comment:
Been thinking about this, could create a `variantsUpdateScript` which consumes the information from `versions.nix`, as part of `mkPassthruAttrs`, we can add a `passthru.variantAttr`, so the update script know which variant is being targeted. Still need to filter out things like `.full` variants, as they aren't really updated so much as they are buildtime descriptions.